From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Akanksha Hotta <akanksha.hotta@intel.com>
Date: Tue, 21 Jul 2026 11:07:45 +0530
Subject: drm/buddy: provide fallback implementation for gpu_buddy on
 older kernels

The buddy allocator core uses gpu_buddy api, where in older kernels they
split full allocator of drm_buddy structure which still resides inside
the itself in header, which was failing when we directly map drm_buddy
to gpu_buddy.

As headers share the __DRM_BUDDY_H__ include guard. Since the
backport include path is searched first, the print-only overlay the kernel
header, leaving struct drm_buddy undefined and gpu_buddy structure incomplete.

Signed-off-by: Rachapalli Bandi Jagadeesh <jagadeesh.rachapalli.bandi@intel.com>
Signed-off-by: Akanksha Hotta <akanksha.hotta@intel.com>
---
 include/drm/drm_buddy.h | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/include/drm/drm_buddy.h b/include/drm/drm_buddy.h
index 3054369..8e2a2fc 100644
--- a/include/drm/drm_buddy.h
+++ b/include/drm/drm_buddy.h
@@ -4,6 +4,7 @@
  */
 
 #ifndef __DRM_BUDDY_H__
+#if defined(HAVE_LINUX_GPU_BUDDY_H_AVAILABLE) || defined(CPTCFG_BUILD_XE_DRM_BUDDY)
 #define __DRM_BUDDY_H__
 
 #include <linux/gpu_buddy.h>
@@ -15,4 +16,9 @@ void drm_buddy_print(struct gpu_buddy *mm, struct drm_printer *p);
 void drm_buddy_block_print(struct gpu_buddy *mm,
 			   struct gpu_buddy_block *block,
 			   struct drm_printer *p);
+
+#else
+#include_next <drm/drm_buddy.h>
+#endif /* HAVE_LINUX_GPU_BUDDY_H_AVAILABLE || CPTCFG_BUILD_XE_DRM_BUDDY */
+
 #endif
-- 
2.43.0

